cachi2/core/package_managers/general.py

```diff
@@ -33,6 +44,172 @@ def download_binary_file(url, download_path, auth=None, insecure=False, chunk_si
+                repo.git.checkout(repo_commit)
+
+                log.info(f"Create tar archive for - {repo_name}")
+                with tarfile.open(os.path.join(download_dir, f"{repo_name}.tgz"), "w:gz") as tar:
+                    tar.add(tmpdir, arcname="package")
+
+    except Exception as exception:
+        log.error(f"Unsuccessful cloning or packing for {parse_url[0]}")
+        raise FetchError(
+            (
+                f"Could not process {parse_url[0]}. "
+                f"exception_name: {exception.__class__.__name__}, "
+                f"details: {exception}"
+            )
+        ) from None
+
+
+async def async_download_binary_file(
+    session: aiohttp_retry.RetryClient,
+    url: str,
+    download_dir: Path,
+    tarball_name: str,
+    auth: Optional[aiohttp.BasicAuth] = None,
+    chunk_size: int = 8192,
+):
+    """
+    Download a binary file (such as a TAR archive) from a URL using asyncio.
+
+    :param aiohttp_retry.RetryClient session: Aiohttp interface for making HTTP requests.
+    :param str url: URL for file download
+    :param str download_dir: Path to download file to
+    :param str tarball_name: Name of the file
+    :param aiohttp.BasicAuth auth: Authentication for the URL
+    :param int chunk_size: Chunk size param for Response.content.read()
+    :raise FetchError: If download failed
+    """
+    try:
+        log.debug(f"Download started - {tarball_name}")
+        async with session.get(url, auth=auth, raise_for_status=True) as resp:
+            with open(os.path.join(download_dir, tarball_name), "wb") as f:
+                while True:
+                    chunk = await resp.content.read(chunk_size)
+                    if not chunk:
+                        break
+                    f.write(chunk)
+
+    except Exception as exception:
+        log.error(f"Unsuccessful download: {tarball_name}")
+        # "from None" since we have the exception context in the logs
+        raise FetchError(
+            (
+                f"Could not download {tarball_name} from {url}. "
+                f"exception_name: {exception.__class__.__name__}, "
+                f"details: {exception}"
+            )
+        ) from None
+
+    log.debug(f"Download completed - {tarball_name}")
+
+
+async def get_dependencies(
+    download_dir: Path,
+    deps_to_download: List[str],
+    concurrency_limit: int,
+) -> List[str]:
+    """
+    Asynchronous function that execute the dependencies download.
+
+    Receives the url (proxy_repo_url), the destination directory (download_dir)
+    and the dependencies to be downloaded (deps_to_download).
+
+    :param Path download_dir: Path to download file to.
+    :param list[str] deps_to_download: List of dependencies to be downloaded.
+    :param int concurrency_limit: Max number of concurrent tasks (downloads).
+    :return: a list of the downloaded tarballs.
+    :rtype: list[str]
+    """
+
+    attempts = 3
```

Comment:
Can we use the same number of retries that we use for synchronous downloads?

https://github.com/containerbuildsystem/cachi2/blob/a1944191430efe2fe158f3518017abe58c2a665a/cachi2/core/http_requests.py#L16-L22

Reply:
I imported DEFAULT_RETRY_OPTIONS, acceptable?

Reply:
One day we could put a `get_aiohttp_session()` method in http_requests.py, but until then this is fine